flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/RowtimeValidator.java

```diff
@@ -0,0 +1,175 @@
+			timestampsValidation.put(ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD, timestampExistingField);
+			timestampsValidation.put(ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_SOURCE, DescriptorProperties.noValidation());
+			timestampsValidation.put(ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM, timestampCustom);
+		} else {
+			timestampsValidation.put(ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD, timestampExistingField);
+			timestampsValidation.put(ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM, timestampCustom);
+		}
+
+		properties.validateEnum(prefix + ROWTIME_TIMESTAMPS_TYPE, false, timestampsValidation);
+
+		Consumer<String> watermarkPeriodicBounded =
+				s -> properties.validateLong(prefix + ROWTIME_WATERMARKS_DELAY, false, 0);
+
+		Consumer<String> watermarkCustom = s -> {
+			properties.validateString(prefix + ROWTIME_WATERMARKS_CLASS, false, 1);
+			properties.validateString(prefix + ROWTIME_WATERMARKS_SERIALIZED, false, 1);
+		};
+
+		Map<String, Consumer<String>> watermarksValidation = new HashMap<>();
+		if (supportsSourceWatermarks) {
+			watermarksValidation.put(ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_ASCENDING, DescriptorProperties.noValidation());
+			watermarksValidation.put(ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_BOUNDED, watermarkPeriodicBounded);
+			watermarksValidation.put(ROWTIME_WATERMARKS_TYPE_VALUE_FROM_SOURCE, DescriptorProperties.noValidation());
+			watermarksValidation.put(ROWTIME_WATERMARKS_TYPE_VALUE_CUSTOM, watermarkCustom);
+		} else {
+			watermarksValidation.put(ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_ASCENDING, DescriptorProperties.noValidation());
+			watermarksValidation.put(ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_BOUNDED, watermarkPeriodicBounded);
+			watermarksValidation.put(ROWTIME_WATERMARKS_TYPE_VALUE_CUSTOM, watermarkCustom);
+		}
+
+		properties.validateEnum(prefix + ROWTIME_WATERMARKS_TYPE, false, watermarksValidation);
+	}
+
+	// utilities
+
+	public static Optional<Tuple2<TimestampExtractor, WatermarkStrategy>> getRowtimeComponents(
+			DescriptorProperties properties, String prefix) {
+		// create timestamp extractor
+		TimestampExtractor extractor;
+		Optional<String> t = properties.getOptionalString(prefix + ROWTIME_TIMESTAMPS_TYPE);
+		if (!t.isPresent()) {
+			return Optional.empty();
+		}
+
+		switch (t.get()) {
+			case ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD:
+				String field = properties.getString(prefix + ROWTIME_TIMESTAMPS_FROM);
+				extractor = new ExistingField(field);
+				break;
+			case ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_SOURCE:
+				extractor = StreamRecordTimestamp.INSTANCE;
+				break;
+			case ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM:
+				Class<TimestampExtractor> clazz = properties.getClass(
+						prefix + ROWTIME_TIMESTAMPS_CLASS, TimestampExtractor.class);
+				extractor = EncodingUtils.decodeStringToObject(
+						properties.getString(prefix + ROWTIME_TIMESTAMPS_SERIALIZED),
+						clazz);
+				break;
+			default:
+				throw new RuntimeException("Unsupported rowtime timestamps type: " + t.get());
```

Comment:
nit: use `ValidationException` here and at other places where the user did something wrong